### ISSUE TITLE
WriteBatch.set should have a signature override

### ISSUE THREAD

**Issue by sk-** [googleapis/nodejs-firestore#1000]
`WriteBatch.set` changes its semantics when passing options, specially when providing `{merge: true}`

In that case an overload to the signature should be added, as in that case there's no need to provide a full document.

The [original definition](c89546f5ae83da3845076aeeffcda75f9b208f5c) is:

```ts
set<T>(documentRef: DocumentReference<T>, data: T, options?: SetOptions): WriteBatch;
``` 

The proposed overload would be something like:
```ts
set<T>(documentRef: DocumentReference<T>, data: Partial<T>, {merge: true}): WriteBatch;
```

**Comment by schmidt-sebastian** [comment]
@sk Oh, that makes a lot of sense, and it doesn't seem like a breaking change. We will get this done shortly.

**Comment by schmidt-sebastian** [comment]
@thebrianchen Brian, can you fix this here and in the Web SDK?

**Comment by sk-** [comment]
This applies to `DocumentReference.set`, `Transaction.set`, and `WriteBatch.set`

**Comment by thebrianchen** [comment]
@sk- After looking into this further, we realized that accepting an overload for `Partial<T>` in `set()` would require modifying the signature of `FirestoreDataConverter.toFirestore()` to also take `Partial<T>`, since Firestore needs to first convert user data into `DocumentData` ([code](https://github.com/googleapis/nodejs-firestore/blob/19c2c75d86c3aab967d21da16598016185ae360b/dev/src/write-batch.ts#L299)).

Expanding the signature of `toFirestore()` to `Partial<T>` would be a breaking change. We're going to look into other options, as well as what this means for the Web SDK. If you have any suggestions, please join in :)

**Comment by sk-** [comment]
@thebrianchen actually the types for `set` and `toFirestore` are currently not correct, as they do not allow to specify [`FieldValue`](https://firebase.google.com/docs/reference/js/firebase.firestore.FieldValue)s.

Also note that currently if you are using a converter, you cannot specify partial objects with either `{merge: true}` or `{mergeFields: [...]}`, and you can't use `FieldValue`s.

Using something like https://pirix-gh.github.io/ts-toolbelt/modules/_object_update_.html, would allow to update the fields.

```ts
import {Any, Union} from 'ts-toolbelt';
import {FieldValue} from '@google-cloud/firestore'; 

export type UpdateAll<O extends object, A extends any> = {
  [P in keyof O]: Union.Replace<A, Any.x, O[P]>;
} & {};

export type AddFieldValues<O extends object> = UpdateAll<O, Any.x | FieldValue>

interface Data {
  foo: string;
  bar: number;
}

type UpdateData = AddFieldValues<Data>;
```

So `set` should accept `AddFieldValues<T>` or `Partial<AddFieldValues<T>>` depending on the args.

I understand is tricky to decide what to do with the converter. I think though, that at the very least it should be updated so that `toFirestore` can accept data with `FieldValue`s. With respect to optional fields you could add a new `toFirestoreWithOptionals` (bike-shedding needed) optional method, which is only used if you are merging docs, and raising an error if the method is not defined. The same could be done to support `FieldValue`s.

**Comment by thebrianchen** [comment]
@sk- Can you elaborate on your use case of how you're passing in FieldValues? I trying to understand why we would need an additional type to support FieldValues.

`toFirestore()` only checks that the user provided objected can be converted into valid `DocumentData`. Once we've added support for `Partial`, you should be able to pass in FieldValues with `mergeFields`, so you could just do:
```
interface Post {
  name: string;
  author: string;
  views: number;
};
 
const post = await doc.get();
post.views = FieldValue.increment(1);
writeBatch.set(post, {mergeFields: ['views']});
```

**Comment by thebrianchen** [comment]
@sk- Can you elaborate on your use case of how you're passing in FieldValues? I follow along with the need for `Partial` to support `merge` and `mergeFields`, but I'm trying to understand why we would need an additional type to support FieldValues.

`toFirestore()` only checks that the user provided objected can be converted into valid `DocumentData`. As long as it's valid `DocumentData`, you can modify `toConverter` to fit your needs. The following example illustrates how you could use FieldValues with `merge` once it's supported:
```
interface Post {
  name: string;
  author: string;
  createdAt: number;
};

const postConverter = {  
  toFirestore(post: Partial<Post>): FirebaseFirestore.DocumentData {
    return {
      title: post.title,
      author: post.author,
      lastModified: FirebaseFirestore.FieldValue.serverTimestamp()
    };
 },
 fromFirestore(
   data: FirebaseFirestore.DocumentData
 ): Post {
   return new Post(data.title, data.author, data.lastModified);
 }
};
 
const post = updatePostWithData();
writeBatch.set(post, {merge: true});
```

**Comment by sk-** [comment]
@thebrianchen you are right about the converter only needs to generate a `DocumentaData`, however if `post` in your example has any `FieldValue`s then it won't be compatible with the `Post` type. That applies both to the calls to `create`, `update`, as well as the converter call `toFirestore`.

Consider how would you append a value to an array without using `FieldValue`s directly in the document.

Check this [TS Playground Link](https://www.typescriptlang.org/play/#code/MYGwhgzhAEBiCWBTEATAamEBXR0De0AvgFDHwB2ALogE4BmYwuACgPYSX7THS-SXxKIRAC5oHGhQDmAGh58wWSgAtWNMROly+0cBwCyrFPDpIUYgCJhqcksWGcADu0pi2HaAF5883gKGi0ADkQdoKSqrqwaG+upCUhsamiObQ5IgA7nBIqBjYiAAUAJTEJEA)

**Comment by richnicholls404** [comment]
This seems closely related to an issue I was about to create which not only affects `set()`, but also `update()` (both the single-field and object implementations). `update(data: UpdateData, ...)` would benefit from being  `update(data: Partial<T>, ...)` (this repeats 2 of 3 times in type definitions file). Again, this would cause the issues you are discussing in the data converter.

Here is a code sample to show what I mean. Read the comments, mostly ignore the bootstrapping code :) : https://repl.it/@RichNicholls/UnequaledGuiltyGlueware

And cheers guys, I love this library!

**Comment by thebrianchen** [comment]
@sk- 

> Consider how would you append a value to an array without using FieldValues directly in the document.

If the original `Post` object was not typed to accept `FieldValue`, how would you pass it in the object to Firestore to begin with? Assuming you have an existing `Post` object and want to add a server timestamp, wouldn't you have to add it to the original object before passing it in? Either way, you would have to rewrite them in the converter or include them in your class.
```
let post: Post = {
    title: '',
    author: '',
    lastModified: Date.now()
};

// Requires your object type to support FieldValue.
post.lastModified = new FieldValue.ServerTimestamp();  
writeBatch.set(post);
```
 
Can you provide a code snippet with how you could use Firestore with FieldValues if we implemented the changes mentioned in your [comment](https://github.com/googleapis/nodejs-firestore/issues/1000#issuecomment-612448524)?

**Comment by thebrianchen** [comment]
@richnicholls404 We cannot add `Partial` support for `update()` because the input parameter for `update()` is `UpdateData` ([documentation](https://googleapis.dev/nodejs/firestore/latest/global.html#UpdateData)), which is a mapping from field paths to values, rather than fields to values.

**Comment by sk-** [comment]
@thebrianchen Update paths does accept fields (strings), not only field paths. By using something like [`Paths`](https://pirix-gh.github.io/ts-toolbelt/modules/_object_paths_.html) or [`Flatten`](https://medium.com/@flut1/deep-flatten-typescript-types-with-finite-recursion-cb79233d93ca), you could generate the actual paths.

So `UpdateData` could be something like:

```ts
Partial<AddFieldValues<T>> & Partial<AddFieldValues<Flatten<T>>> & Record<FieldPath, any>
```

With converters things get more complicated and we are not using them in our team at the moment, so can't talk much about that.

However, for the case of `set` and using interfaces for the data model, my proposal works perfectly.

```ts
interface Post {
  title: string;
  author: string;
  lastModified: firestore.Timestamp;
}

const post = {
  title: updatedTitle,
  lastModified: firestore.FieldValue.serverTimestamp(),
}

ref.set(post, {merge: true})
```

The call to `set` would work, because the data type for data would be `Partial<AddFieldValues<Post>>`, which basically would be:

```
interface OptionalPostWithFieldValues {
  title?: string;
  author?: string;
  lastModified?: firestore.Timestamp | firestore.FieldValue;
}
``` 

adding `FieldValue` to the list of types could be made to be more type safe if there were multiple subclasses. For example, if the type of the property is array you should only add `FieldValue.arrayUnion` and `FieldValue.arrayRemove`, but not  `FieldValue.increment`.

**Comment by thebrianchen** [comment]
**Regarding supporting `Partial<T>` support for `update()`:** 
I see what you mean about how we can flatten `Partial<T>` into `UpdateData`.  `update()` currently accepts `field: string` arguments that alternate between field paths and field values, and the introduction of passing in nested objects as `Partial<T>` would expand the API surface in a non-intuitive direction. There is currently substantial overlap between `set()` with `merge` and `update()`, and adding this additional overload to `update()` would further blur the differences. 

**Regarding support for `FieldValue`:**
Thanks for taking the time to write up the use case -- that's a really cool use of ts-toolbelt! However, after talking with the team, we concluded that such a change would add a non-trivial degree of complexity to the SDK in terms of overloads and additional types. Considering that we already support this behavior without converters, we've decided to not prioritize this unless there is substantial developer demand for this feature.

I am currently working on an adding an optional method to `FirestoreDataConverter` that would support passing in `Partial<T>` into `set()` methods (#1060). Thanks so much for filing the feature request and writing detailed explanations!
